File: Tree/C/M814-binary-tree-pruning.c
// time:  O(n)
// space: O(n)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* pruneTree(struct TreeNode* root) {
    if(root->left)
        root->left = pruneTree(root->left);
    
    if(root->right)
        root->right = pruneTree(root->right);
    
    if(root->val == 0 && !root->left && !root->right)
        return NULL;
    
    return root;
}

File: Tree/Python3/M814-binary-tree-pruning.py
# time:  O(n)
# space: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pruneTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if root.left:
            root.left = self.pruneTree(root.left)
            
        if root.right:
            root.right = self.pruneTree(root.right)
            
        if root.val == 0 and not root.right and not root.left:
            return None
        return root